Proxy sharding

mcb30 opened the pull request:
Add support for sharding requests across multiple proxies using consistent hashing, controlled via the `CVMFS_PROXY_SHARD` option in the client configuration.  This allows load to be distributed approximately equally across multiple proxies, with clients independently choosing the same proxy for a given piece of content.

This pull request does not implement bounded loads (as used in e.g. haproxy's `hash-type consistent`) since there is no obvious way for clients to share knowledge about pending requests, and there is probably limited utility in bounding the per-proxy load generated by a single client acting in isolation.

mcb30 committed:
Allow ValidateProxyIpsUnlocked() to indicate potential proxy changes

With multiple concurrent proxies in use, the process of selecting a
proxy is potentially expensive.  Allow ValidateProxyIpsUnlocked() to
indicate whether or not the proxy list may have changed, to avoid the
need to always reselect a proxy after verifying that the DNS records
are up to date.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Allow for proxy to be selected on a per-transfer basis

Allow for a proxy selection mechanism that can choose between any of
the currently non-burned proxies for a particular transfer.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Provide a partial hash digest for use as an unsigned integer key

The existing digest algorithms satisfy the avalanche criteria and so
any partial subset (e.g. the first 32 bits of the digest result) may
sensibly be used as a key in contexts where collisions are unimportant
(e.g. for load-balancing).

Provide a helper method that extracts the first 32 bits of the digest
result as an unsigned 32-bit integer (in an endian-independent way).

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Allow for requests to be sharded across multiple proxies

Implement proxy sharding via consistent hashing, with the proxy hash
keys calculated from the (DNS-resolved) URLs and the request hash keys
derived from the already-known content hash.

Handle the single-proxy case by building a map that contains only a
single entry, thereby allowing the new code paths to match the current
behaviour.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Define CVMFS_PROXY_SHARD configuration option

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Run proxy failover tests in both single-proxy and sharded modes

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

jblomer commented:
@cernvm-bot cloudtest

jblomer commented:
@cernvm-bot unittest

jblomer commented:
@cernvm-bot cpplint

jblomer commented:
@cernvm-bot cpplint

jblomer commented:
@cernvm-bot tidy

jblomer commented:
@mcb30 Many thanks, that looks super interesting!  We are currently cutting the 2.9 release, so we'll schedule the sharding feature for the next one.  Give us a few days for an thorough review.

mcb30 commented:
Pushed an update to fix the linter errors.  Will review the build errors as soon as I'm able to view them.

mcb30 commented:
> building for cloudtests finished: FAILURE https://lcgapp-services.cern.ch/cvmfs-jenkins/job/CvmfsFullBuildDocker/3114/

Something seems to be broken in the Mac build environment:
```
  CC       libcares_la-ares__close_sockets.lo
In file included from ares__close_sockets.c:17:
In file included from ./ares_setup.h:214:
In file included from ./setup_once.h:33:
In file included from /usr/local/include/stdio.h:64:
/usr/local/include/_stdio.h:93:16: warning: pointer is missing a nullability type specifier (_Nonnull, _Nullable, or _Null_unspecified) [-Wnullability-completeness]
        unsigned char   *_base;
                        ^
/usr/local/include/_stdio.h:93:16: note: insert '_Nullable' if the pointer may be null
        unsigned char   *_base;
                        ^
                          _Nullable 
/usr/local/include/_stdio.h:93:16: note: insert '_Nonnull' if the pointer should never be null
        unsigned char   *_base;
                        ^
                          _Nonnull 
/usr/local/include/_stdio.h:138:32: warning: pointer is missing a nullability type specifier (_Nonnull, _Nullable, or _Null_unspecified) [-Wnullability-completeness]
```
with thousands of similar errors all seemingly related to `-Wnullability-completeness`

jblomer commented:
The warnings are cumbersome but should be a result of us building some externals ourselves.  The actual error seems to be this:

    /Users/sftnight/jenkins/workspace/CvmfsFullBuildDocker/CVMFS_BUILD_ARCH/osx10-x86_64/CVMFS_BUILD_PLATFORM/mac/cvmfs/cvmfs/download.cc:2624:31: error: expected expression
            opt_proxy_map_.insert({key, proxy});
                                  ^
    /Users/sftnight/jenkins/workspace/CvmfsFullBuildDocker/CVMFS_BUILD_ARCH/osx10-x86_64/CVMFS_BUILD_PLATFORM/mac/cvmfs/cvmfs/download.cc:2629:27: error: expected expression
        opt_proxy_map_.insert({(uint32_t)-1, opt_proxy_map_.rbegin()->second});
                              ^
    /Users/sftnight/jenkins/workspace/CvmfsFullBuildDocker/CVMFS_BUILD_ARCH/osx10-x86_64/CVMFS_BUILD_PLATFORM/mac/cvmfs/cvmfs/download.cc:2634:27: error: expected expression
        opt_proxy_map_.insert({(uint32_t)-1, proxy});

DrDaveD commented:
@mcb30 Have you done any measurements that demonstrate how much this improves performance?

mcb30 commented:
> @mcb30 Have you done any measurements that demonstrate how much this improves performance?

Not specifically for this implementation.  The overall approach of using consistent hashing is a very well-known design that improves cache locality when multiple caching proxies are in use, and the overall impact on system performance is going to depend heavily on the cache system design and implementation.

For us, this will be eliminating a separate sharding layer that is currently implemented using haproxy, and the impact will therefore be reduced bandwidth and CPU consumption since traffic does not first need to pass through an haproxy instance before reaching the real caching proxy.

The cost of performing the balanced tree lookup on each request is negligible in comparison to any download time, and so I would not expect to be able to measure any impact from this extra code path.  The (re)construction of the balanced tree happens only in the event of a proxy configuration change or proxy failure, which is sufficiently rare to make the cost also negligible.

Did you have a specific concern, or was it just general curiosity about the performance improvement?

DrDaveD commented:
I was just wondering if it was really worth the added complexity.   My own sense has been that the cache hit rates in our applications are typically so high that added delays from downloading the same object to a few more caches is negligible.  We also usually have plenty of disk space on the caches -- not a lot of space is needed -- so I don't see much benefit from the cache space savings.

mcb30 commented:
> I was just wondering if it was really worth the added complexity.

Most of the PR deals with cleaning up the existing code, and I'd suggest merging it even without the ability to shard across multiple proxies.  The extra code that implements the sharding logic is isolated to about 20 lines of code in `UpdateProxiesUnlocked()`: I've deliberately kept it clean and simple.

> My own sense has been that the cache hit rates in our applications are typically so high that added delays from downloading the same object to a few more caches is negligible. We also usually have plenty of disk space on the caches -- not a lot of space is needed -- so I don't see much benefit from the cache space savings.

We definitely need the sharding with the size of data sets that we use, which is why we currently have the haproxy front end.

mcb30 commented:
> The warnings are cumbersome but should be a result of us building some externals ourselves. The actual error seems to be this:
> 
> ```
> /Users/sftnight/jenkins/workspace/CvmfsFullBuildDocker/CVMFS_BUILD_ARCH/osx10-x86_64/CVMFS_BUILD_PLATFORM/mac/cvmfs/cvmfs/download.cc:2624:31: error: expected expression
>         opt_proxy_map_.insert({key, proxy});
>                               ^
> /Users/sftnight/jenkins/workspace/CvmfsFullBuildDocker/CVMFS_BUILD_ARCH/osx10-x86_64/CVMFS_BUILD_PLATFORM/mac/cvmfs/cvmfs/download.cc:2629:27: error: expected expression
>     opt_proxy_map_.insert({(uint32_t)-1, opt_proxy_map_.rbegin()->second});
>                           ^
> /Users/sftnight/jenkins/workspace/CvmfsFullBuildDocker/CVMFS_BUILD_ARCH/osx10-x86_64/CVMFS_BUILD_PLATFORM/mac/cvmfs/cvmfs/download.cc:2634:27: error: expected expression
>     opt_proxy_map_.insert({(uint32_t)-1, proxy});
> ```

@jblomer Thanks for pointing this out.  I've updated the PR to avoid using `std::map` literals, and rebased onto the latest `origin/devel`.

jblomer commented:
> I was just wondering if it was really worth the added complexity. My own sense has been that the cache hit rates in our applications are typically so high that added delays from downloading the same object to a few more caches is negligible. We also usually have plenty of disk space on the caches -- not a lot of space is needed -- so I don't see much benefit from the cache space savings.

I see the benefit for software in improved cold-cache behavior, as it is enough to populate one of the possibly many caches with a new object. The space savings are relevant more for conditions data than for software.

jblomer commented:
@cernvm-bot cloudtest

mcb30 commented:
> cloudtests finished: FAILURE https://lcgapp-services.cern.ch/cvmfs-jenkins/job/CvmfsCloudTesting/1305/

Looks like a completely unrelated failure in a single test that doesn't use a proxy in any way (as far as I can tell):
```
*** NOTE: This is currently an experimental CernVM-FS feature

Entering ephemeral writable shell for /cvmfs/test.repo.org... 
Mounting CernVM-FS read-only layer... (unexpected termination) cannot mount cvmfs read-only branch (16)
  command: `/usr/bin/cvmfs2 -o allow_other,config=/home/sftnight/.cvmfs/test.repo.org/session.EIRse2/client.config test.repo.org /home/sftnight/.cvmfs/test.repo.org/session.EIRse2/rdonly`

Stacktrace:
/usr/bin/cvmfs_publish(+0x46d61) [0x556e717dad61]
/usr/bin/cvmfs_publish(+0x33875) [0x556e717c7875]
/usr/bin/cvmfs_publish(+0x392d2) [0x556e717cd2d2]
/usr/bin/cvmfs_publish(+0x3bdf3) [0x556e717cfdf3]
/usr/bin/cvmfs_publish(+0x15828) [0x556e717a9828]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f5961a090b3]
/usr/bin/cvmfs_publish(+0x15a3e) [0x556e717a9a3e]
```

mcb30 commented:
Rebased to resolve merge conflict

radupopescu commented:
Thanks for the contribution, @mcb30!

Looks good, merging.
